### ISSUE TITLE
[Xcode 12][SwiftPM] App fails to build when linking to AppsFlyer 6.1.0 or fails to Export for AppsFlyer 6.1.2 

### ISSUE THREAD

**Issue by Panajev** [AppsFlyerSDK/AppsFlyerFramework#143]
# Report

## SDK Version

6.1.0 - 6.1.2 (former fails to codesign, latter fails to export)

## What did you do?

Replaced the AppsFlyer CocoaPod pod with the SPM version

## What did you expect to happen?

Application to compile successfully

## What happened instead?

App fails to build when linking to AppsFlyer 6.1.0 code signing is invoked for it and fails

## Please provide any other relevant information.

Hello, we understand the SPM version suggested for AppsFlyer is 6.0.8, but does it mean that it is SPM supported from that version onwards or just 6.0.8 and not 6.1.0? Can documentation be improved to reflect this?

**Comment by alextudge** [comment]
When using 6.0.8 via SPM we got an error on run about a missing plist.
Using 6.1.0 via SPM produces a compile time warning as above.

**Comment by akolov** [comment]
I did some poking around xcframework embedded into this package, and apparently codesign doesn't like when there's Versions dir in the framework bundle. Code signing works file if you move Headers and AppsFlyerLib into the bundle root and delete Versions dir.

**Comment by s-hintz** [comment]
AppsFlyerLib.framework is copied twice into our apps.

PlugIns/AppsFlyerLib.framework is incorrect
Frameworks/AppsFlyerLib.framework is correct

So we get this error:
PlugIns/AppsFlyerLib.framework: bundle format unrecognized, invalid, or unsuitable
Command CodeSign failed with a nonzero exit code

This happens in our "Embed App Extensions" build phase.

**Comment by Panajev** [comment]
@s-hintz is it happening with the 6.1.1 release too?

**Comment by af-fess** [comment]
hey @Panajev we are checking this. 
Thank you

**Comment by Panajev** [comment]
Thank you :)

**Comment by s-hintz** [comment]
I checked 6.1.0 and 6.1.1 and I could not get it to work.
I even tried to delete the plugin with a script, but since it I copied and code signed in the same build phase, I was not able to delete it in between.

#!/bin/sh
rm -r $TARGET_BUILD_DIR/$PLUGINS_FOLDER_PATH/AppsFlyerLib.framework

**Comment by s-hintz** [comment]
We are using several Swift Packages.
AppsFlyer is the only one with problems.

**Comment by s-hintz** [comment]
We had to revert our changes and use the 6.1.0 with Carthage now.

**Comment by akolov** [comment]
For those having issues with codesign I made a fork that makes it happy until upstream is fixed: https://github.com/akolov/AppsFlyerFramework.git
Please note it doesn't fix extra Frameworks and Plugins embedding, just codesign fail.

**Comment by ivanlisovyi** [comment]
@s-hintz Duplicated frameworks issue (and code-signing issue as well) is not an AppsFlyer issue rather the SPM issue. The same issue happens with [Firebase](https://github.com/firebase/firebase-ios-sdk/issues/6472#issuecomment-694449182) and other frameworks that use `binaryTarget` in the SPM `Package.swift`. 

To make sure that your script works, you need to make sure to execute it as post-action in your scheme instead of build phase. (See linked Firebase issue for more info)

**Comment by akolov** [comment]
@ivanlisovyi this doesn't solve the issue when AppsFlyer is embedded within another Framework. Code signing happens *after* post-action scripts are run.

**Comment by ivanlisovyi** [comment]
@akolov Yes, it doesn't. It only solves the framework duplication issue which AppsFlyers won't be able to solve anyways unlike the code signing.

**Comment by andr-ggn** [comment]
Really appreciate to:
@akolov 
@ivanlisovyi 
We updated master branch to have flatten MacCatalyst framework. So it should solve `code-signing issue`. Next version of SDK will contain flatten xcframework.

**Comment by quentinfasquel** [comment]
@andr-ggn I am still getting a signing issue with version 6.1.1. I am using Swift Package Manager and have a custom package, let say _CusomTracker_, that brings the AppsFlyer dependency, and my final app uses _CustomTracker_ but fails at the last step when about to sign AppsFlyerLib.framework. This may be an important note : the app is iOS only, no iPad no Mac Catalyst.

I would surely appreciate a solution.

**Comment by quentinfasquel** [comment]
In fact @andr-ggn the issue is actually obvious. It seems you guys changed delivering a dynamic framework for a static framework. With SPM & xcframeworks, from my understanding Xcode will only accept either a static **library** or a dynamic framework.

And the reason I am saying that you are delivering a static framework is only because of its structure, with "Versions", Current/A

**Comment by akolov** [comment]
@quentinfasquel I believe the fix was not included in 6.1.1 yet, try switching SPM to use head or the master branch instead.

**Comment by quentinfasquel** [comment]
@akolov indeed ! You should tag that immediately ;)

**Comment by quentinfasquel** [comment]
@andr-ggn / @akolov it was too quick to be right, two major issues.

TLDR: 1- If you intend to distribue it as a framework, it's gonna have to be dynamic, if you just wanna ship a binary with SPM, it has to be through a xcframework, which can only support a **static library** or a **dynamic framework**. No damn f*ing static framework. 2- tag the fix quickly.

- First, not having the tag and branching on master forces every intermediate swift package to be brought through branch instead of tag. (I had CustomTracker with tag x.x.x bringing AppsFlyer on branch master, then on revision... SPM cannot resolve a version if it depends on an unversioned dependency)

- Second, and most critical. I just tried archiving, it works. But exporting the archive results in the following issue :
**Error Domain=DVTMachOErrorDomain Code=0 "Found an unexpected Mach-O header code: ...**

The issue you guys previously fixed was, I am guessing, was : "flatten the framework so that Xcode can sign it, or so that SPM stops complaining". The truth is, it should not have tried to sign it in the first place.

What's happening now is that Xcode/SPM is being tricked trying to copy a dynamic framework, but in the end it's a static framework. The concept of static framework doesn't really exist, especially not in the **xcframework** world. It used to be a hack until iOS apps could support actual dynamic frameworks. The issue appears now later on in the distribution flow,  `exportArchive` complains because it doesn't know what to do with an `ar archive` (static).

Someone please tell me if I'm wrong in any of this, and please let me know if you've got a fix.

**Comment by akolov** [comment]
`Found an unexpected Mach-O header code` is most likely because frameworks are copied into other embedded frameworks and plugins as per this issue: https://github.com/firebase/firebase-ios-sdk/issues/6472#issuecomment-694449182
I have AppsFlyer currently integrated in the production app in the AppStore via current master branch and it works fine.

**Comment by quentinfasquel** [comment]
@akolov are you adding AppsFlyer with SPM ? this is the issue. static frameworks don't exist and it indeed copies any found framework. the solution therefore is to declare it as what it really is a static library. My solution here https://github.com/quentinfasquel/AppsFlyerFramework and it works fine with SPM all the way through export.

**Comment by akolov** [comment]
@quentinfasquel yes, I'm using it via SPM.

**Comment by quentinfasquel** [comment]
@akolov are you using Xcode 12.2 ?

**Comment by akolov** [comment]
@quentinfasquel yes, Version 12.2 (12B45b).

**Comment by quentinfasquel** [comment]
@akolov I really don't get it then. Brand new project, add AppsFlyerLib as a SPM, branch master as you pointed out earlier in the day. Archive, distribute, export as Ad-Hoc... and Xcode shows this. In logs, it is just the Mach-O issue
<img width="726" alt="Screenshot 2020-11-25 at 17 55 08" src="https://user-images.githubusercontent.com/545401/100258431-5e86ad00-2f47-11eb-8f4b-89dbfb7b9db0.png">

**Comment by akolov** [comment]
@quentinfasquel I'm linking all SPM packages into an extra framework that's shared among several targets in my project. Additionally, I have a cleanup build phase, that remove anything unexpected in embedded Frameworks and Plugins.
This is not just AppsFlyer issue, but Firebase, Stripe and many others are affected as well.

**Comment by quentinfasquel** [comment]
@akolov got it, in the mean time I have a different solution that doesn't require any intermediate custom framework nor any cleanup. Just assuming that a static framework isn't something that exist.

**Comment by joshuapoq** [comment]
Any update on getting master tagged? I've not actually tried it but I assume it resolved the codesigning issue with Xcode 12.2 - we can't use master as it goes against how SPM using versions as stable releases works.

Check the [SPM proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0150-package-manager-branch-support.md#specifying-branches-or-revisions-in-the-manifest) that states this:
>While this feature [specifying branches] is useful during development, a package’s dependencies should be updated to point at versions instead of branches before that package is tagged for release. This is because a released package should provide a stable specification of its dependencies, and not break when a branch changes over time. **To enforce this, it is an error if a package referenced by a version-based dependency specifies a branch in any of its dependencies.**

**Comment by bettola** [comment]
We really need an update on this one. So unprofessional by AppsFlyer, it's the only remaining framework we have still to add using Pods 🤦🤦🤦

**Comment by andr-ggn** [comment]
Unfortunately, we cannot just add/remove releases/tags.
Each tag is a new release. There is a process in which not only development is involved.
---
6.1.2 with tag

**Comment by andr-ggn** [comment]
Unfortunately, we cannot just add/remove releases/tags.
Each tag is a new release. There is a process in which not only development is involved.

6.1.2 with tag

**Comment by Panajev** [comment]
I have set the version of the framework to 6.1.2 and I am still getting the **Found an unexpected Mach-O header code: 0x72613c21** error :/.

**Comment by ilkerc** [comment]
Same issue on v 6.1.2 using SPM

**Comment by andr-ggn** [comment]
@Panajev, @ilkerc This is another SPM issue. Did you try https://github.com/AppsFlyerSDK/AppsFlyerFramework/issues/142 ?

**Comment by ilkerc** [comment]
@andr-ggn Thanks for the suggestion, was a fix for iPhone target but not for catalyst. Build failed with the following error message;

```
/Users/ilker/Library/Developer/Xcode/DerivedData/hello-hgbirkcajlqormcajetinxjyhjnp/Build/Products/Debug - Staging-maccatalyst/Hello.app/Contents/PlugIns/AppsFlyerLib.framework/Versions/A: bundle format unrecognized, invalid, or unsuitable
Command CodeSign failed with a nonzero exit code
```

I have also added the `Contents` folders, same error as above.

```
rm -rf "${TARGET_BUILD_DIR}/${TARGET_NAME}.app/Frameworks/AppsFlyerLib.framework"
rm -rf "${TARGET_BUILD_DIR}/${TARGET_NAME}.app/Plugins/AppsFlyerLib.framework"
rm -rf "${TARGET_BUILD_DIR}/${TARGET_NAME}.app/Contents/Frameworks/AppsFlyerLib.framework"
rm -rf "${TARGET_BUILD_DIR}/${TARGET_NAME}.app/Contents/Plugins/AppsFlyerLib.framework"
```

**Comment by Panajev** [comment]
Thanks @andr-ggn 6.1.2 and that approach sorted us out!

**Comment by andr-ggn** [comment]
@ilkerc You use not a 6.1.2 version

**Comment by ilkerc** [comment]
@andr-ggn I can verify that SPM fetches version 6.1.2. 

<img width="268" alt="Screen Shot 2020-12-08 at 15 24 44" src="https://user-images.githubusercontent.com/6921843/101483627-81628980-3969-11eb-9ce3-c7917f6394f8.png">

I can build the project for iPhone devices and simulators but fails on mac target.
<img width="1284" alt="Image 2020-12-08 at 3 23 28 PM" src="https://user-images.githubusercontent.com/6921843/101483593-71e34080-3969-11eb-9383-943b0d7bdcc6.png">

I have created a sandbox project here: https://github.com/ilkerc/appsflyer-spm-sandbox

**Comment by andr-ggn** [comment]
@ilkerc 
I was able to compile for `My Mac`
Steps:
- download your appsflyer-spm-sandbox
- open with Xcode Version 12.0 (12A7209)
- build phases -> new run script phase -> duplicate script from Post Action to Run Script 
- move script to the top of phases. To get order like: Dependencies, Run Script, Compile Sources
- get compilation error
- create Pre-Action for Build. Build settings from appsflyer-spm-sandbox. Put `echo "Apple"`
- get compilation error
- change `Apple` to `AAA AAA`
- compilation success 

🤦‍♀️ Whyyyyyy

Can you try this steps and back here?

**Comment by ilkerc** [comment]
@andr-ggn Interesting approach :) Didn't work on my case. Only difference is the Xcode version: 12.2 (12B45b)

**Comment by andr-ggn** [comment]
@ilkerc 
- changed `@main` to `@UIApplicationMain`
- changed `echo "AAA AAA"` to `echo "AAA AAA1"
- succeed 

try to change Pre-Action few times..

**Comment by ilkerc** [comment]
@andr-ggn That didn't work on my case too, but discovered an interesting solution;

1. Compile with AppsFlyerLib on frameworks.
![Uploading Screen Shot 2020-12-08 at 17.22.40.png…]()

2. Sign error
3. Remove AppsflyerLib from frameworks
<img width="638" alt="Screen Shot 2020-12-08 at 17 23 20" src="https://user-images.githubusercontent.com/6921843/101495831-12416100-397a-11eb-8f3f-b2b46468b7b2.png">

4. Successful sign

As long as you don't clean build folder, signs successfully. I have purged all pre-post run scripts too. Also, I just tried pod 'AppsFlyerFramework-MacCatalyst'. This too fails with the same error, and I can also solve it with this weird solution. 

Obviously this won't work if you try to archive for a release, I'll try to remove frameworks and continue signing manually on terminal.

**Comment by andr-ggn** [comment]
If apple does not fix the static framework for .xcframework in the near future, we will have to switch to dynamic framework

**Comment by alesp56** [comment]
We are still unsuccessful with the .ipa export (archiving works) when we use SPM (Xcode 12.3, export only for iPhone & iPad). We just replaced the previously linked static framework with the SPM, and now we are getting an IPA processing failed error. We tried to add that post-build action. Unsuccessfull. AppsFlyerLib.framework is present only on the root of the *.app package (not in Frameworks/Plugins folders). This should be ok, right?

**Comment by quentinfasquel** [comment]
> AppsFlyerLib.framework is present only on the root of the *.app package (not in Frameworks/Plugins folders). This should be ok, right?

@alesp56 nope. It doesn't matter where AppsFlyerLib.framework is in the .app package, the app is already linked with it since its a static framework. It must be removed from the .app package. You should do so through a script build phase. This is an Xcode + SPM + static binaries issue... SPM thinks every xcframework contains a dynamic framework and copy it. I'll try and see if there's a generic script that could be used to remove every static binary from .app package.

**Comment by SergeyPetrachkov** [comment]
I do have a problem with 6.1.2 version:
Once I do `import AppsFlyerLib`, the project fails to build with the following error: 

ignoring file /Users/sergeypetrachkov/Library/Developer/Xcode/DerivedData/JuiceFit-etzvubmelzhgcpdhdlxyoibjzllu/Build/Products/Debug-iphonesimulator/AppsFlyerLib.framework/AppsFlyerLib, missing required architecture arm64 in file /Users/sergeypetrachkov/Library/Developer/Xcode/DerivedData/JuiceFit-etzvubmelzhgcpdhdlxyoibjzllu/Build/Products/Debug-iphonesimulator/AppsFlyerLib.framework/AppsFlyerLib (2 slices)
Undefined symbols for architecture arm64:
  "_OBJC_CLASS_$_AppsFlyerLib", referenced from:
      objc-class-ref in AppDelegate.o
ld: symbol(s) not found for architecture arm64

**Comment by quentinfasquel** [comment]
@SergeyPetrachkov I was gonna say, you are on M1.

Not sure there was a tagged version available with SPM. But the MacCatalyst xcframework contains all necessary architectures. Try the same revision ?
`
.package(name: "AppsFlyerLib", url: "https://github.com/AppsFlyerSDK/AppsFlyerFramework.git", .revision("b233d5611aad73f014f07f0440300e7c4db1adb4")),
`

**Comment by SergeyPetrachkov** [comment]
@quentinfasquel  I'll try in a few hours and will report back to you.

**Comment by SergeyPetrachkov** [comment]
@quentinfasquel  nope, still fails to compile

**Comment by SergeyPetrachkov** [comment]
<img width="501" alt="Screenshot 2020-12-17 at 13 14 38" src="https://user-images.githubusercontent.com/7995896/102474545-d38e6380-4069-11eb-90dc-eae6b904900d.png">
this is what I get in derived data

**Comment by bettola** [comment]
Any update? It's appalling that we don't have a fix yet

**Comment by SergeyPetrachkov** [comment]
Ended up doing the following on my M1 Mac:

* AppsFlyer lib added as SPM
* All AppsFlyer calls and import covered by `#if !targetEnvironment(simulator) #endif`
* post build script for my scheme: 
<img width="923" alt="Screenshot 2021-01-17 at 09 17 46" src="https://user-images.githubusercontent.com/7995896/104832790-dff22000-58a4-11eb-8f52-df1a7bd16cb8.png">

And only then it all works and I'm able to build the app for the AppStore.

**Comment by alxrguz** [comment]
Hello, it's been 2 months since the problem was discovered, is there any progress? Apps with AppsFlyer installed via SPM still cannot be uploaded to the AppStore

**Comment by SergeyPetrachkov** [comment]
> Hello, it's been 2 months since the problem was discovered, is there any progress? Apps with AppsFlyer installed via SPM still cannot be uploaded to the AppStore

Hi! I managed to release my app yesterday with Appsflyer added as spm. See my comment above, try doing the same steps?

**Comment by bettola** [comment]
> > Hello, it's been 2 months since the problem was discovered, is there any progress? Apps with AppsFlyer installed via SPM still cannot be uploaded to the AppStore
> 
> Hi! I managed to release my app yesterday with Appsflyer added as spm. See my comment above, try doing the same steps?

Understand there might be a workaround, but we would really love a package who works out of the box, without fiddling with scripts or else; I guess it's fair?

**Comment by SergeyPetrachkov** [comment]
> > > Hello, it's been 2 months since the problem was discovered, is there any progress? Apps with AppsFlyer installed via SPM still cannot be uploaded to the AppStore
> 
> > 
> 
> > Hi! I managed to release my app yesterday with Appsflyer added as spm. See my comment above, try doing the same steps?
> 
> 
> 
> Understand there might be a workaround, but we would really love a package who works out of the box, without fiddling with scripts or else; I guess it's fair?

Sure! I myself would be very glad to get rid of those scripts. But for now it seems to be the only way to proceed with the SPM approach.

**Comment by alxrguz** [comment]
> > Hello, it's been 2 months since the problem was discovered, is there any progress? Apps with AppsFlyer installed via SPM still cannot be uploaded to the AppStore
> 
> Hi! I managed to release my app yesterday with Appsflyer added as spm. See my comment above, try doing the same steps?

Thanks for the recommendation, I'll try tomorrow)

**Comment by Kondamon** [comment]
Export only works for me when I add this as a post-action build phase. (see above)
```
rm -rf "${TARGET_BUILD_DIR}/${PRODUCT_NAME}.app/Frameworks/AppsFlyerLib.framework"
rm -rf "${TARGET_BUILD_DIR}/${PRODUCT_NAME}.app/PlugIns/AppsFlyerLib.framework"
```

**Comment by EmDee** [comment]
According to https://github.com/firebase/firebase-ios-sdk/issues/6472#issuecomment-772589092 the static framework issue has been fixed with Xcode 12.5.

**Comment by kravtsovguy** [comment]
Solving the problem (before Xcode 12.5):

- add build phase run script (after embedding and so on):

```
echo "Removing static frameworks from ${TARGET_NAME}.app"
find "${BUILT_PRODUCTS_DIR}/${TARGET_NAME}.app/" -name '*.framework' -print0 | while IFS= read -r -d '' fm; do
    name=$(basename "${fm}" .framework)
    target="${fm}/${name}"
    echo "Checking: ${fm}"
    if file "${target}" | grep -q "current ar archive"; then
        rm -rf "${fm}"
        echo "Removed static framework: ${fm}"
    fi
done
```
